File: apps/android/core/src/main/java/com/monitor/device/core/api/DeviceApi.kt
package com.monitor.device.core.api

import com.monitor.device.core.model.ChangePasswordRequest
import com.monitor.device.core.model.ChatMediaPage
import com.monitor.device.core.model.ChatMessageDto
import com.monitor.device.core.model.ChatMessagesPage
import com.monitor.device.core.model.ChatSearchPage
import com.monitor.device.core.model.ChatThreadDto
import com.monitor.device.core.model.EmptyJsonBody
import com.monitor.device.core.model.DeviceMeResponse
import com.monitor.device.core.model.DeviceStatusResponse
import com.monitor.device.core.model.DeviceStatusUpdate
import com.monitor.device.core.model.EditChatRequest
import com.monitor.device.core.model.InitUploadRequest
import com.monitor.device.core.model.InitUploadResponse
import com.monitor.device.core.model.LinkDeviceRequest
import com.monitor.device.core.model.LinkDeviceResponse
import com.monitor.device.core.model.LinkedDeviceDto
import com.monitor.device.core.model.OkResponse
import com.monitor.device.core.model.PairingCodeResponse
import com.monitor.device.core.model.PairRequest
import com.monitor.device.core.model.PairResponse
import com.monitor.device.core.model.PairStatusResponse
import com.monitor.device.core.model.PublisherTokenResponse
import com.monitor.device.core.model.PaymentInvoiceDto
import com.monitor.device.core.model.PurchasePlanRequest
import com.monitor.device.core.model.ReactChatRequest
import com.monitor.device.core.model.RegisterPushTokenRequest
import com.monitor.device.core.model.SendChatRequest
import com.monitor.device.core.model.SetCameraFacingRequest
import com.monitor.device.core.model.SubscriptionDto
import com.monitor.device.core.model.SupportSummaryDto
import com.monitor.device.core.model.UpdateProfileRequest
import com.monitor.device.core.model.UploadAvatarRequest
import com.monitor.device.core.model.RecordingListResponse
import com.monitor.device.core.model.RecordingPlaybackRequest
import com.monitor.device.core.model.RecordingPlaybackResponse
import com.monitor.device.core.model.ViewerTokenResponse
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

interface DeviceApi {
    @GET("devices/pair-status")
    suspend fun pairStatus(
        @Query("phone") phone: String? = null,
        @Query("installId") installId: String? = null,
        @Query("signals") signals: String? = null,
    ): PairStatusResponse

    @POST("devices/pair")
    suspend fun pair(@Body request: PairRequest): PairResponse

    @GET("devices/me")
    suspend fun me(
        @Header("Authorization") authorization: String,
    ): DeviceMeResponse

    @PATCH("devices/me")
    suspend fun updateProfile(
        @Header("Authorization") authorization: String,
        @Body body: UpdateProfileRequest,
    ): DeviceMeResponse

    @PATCH("devices/me/password")
    suspend fun changePassword(
        @Header("Authorization") authorization: String,
        @Body body: ChangePasswordRequest,
    ): OkResponse

    @POST("devices/me/avatar")
    suspend fun uploadAvatar(
        @Header("Authorization") authorization: String,
        @Body body: UploadAvatarRequest,
    ): DeviceMeResponse

    @DELETE("devices/me/avatar")
    suspend fun deleteAvatar(
        @Header("Authorization") authorization: String,
    ): DeviceMeResponse

    @GET("devices/me/linked")
    suspend fun linkedDevices(
        @Header("Authorization") authorization: String,
    ): List<LinkedDeviceDto>

    @DELETE("devices/me/linked/{id}")
    suspend fun unlinkDevice(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
    ): OkResponse

    @POST("devices/me/pairing-codes")
    suspend fun createPairingCode(
        @Header("Authorization") authorization: String,
        @Body body: EmptyJsonBody = EmptyJsonBody(),
    ): PairingCodeResponse

    @POST("devices/me/link")
    suspend fun linkDevice(
        @Header("Authorization") authorization: String,
        @Body body: LinkDeviceRequest,
    ): LinkDeviceResponse

    @POST("streaming/devices/{id}/device-viewer-token")
    suspend fun deviceViewerToken(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
    ): ViewerTokenResponse

    @POST("devices/me/linked/{id}/camera")
    suspend fun setLinkedCamera(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Body body: SetCameraFacingRequest,
    ): DeviceStatusResponse

    @GET("recordings/device/{deviceId}")
    suspend fun linkedRecordings(
        @Header("Authorization") authorization: String,
        @Path("deviceId") deviceId: String,
    ): RecordingListResponse

    @POST("recordings/device/{deviceId}/start")
    suspend fun startLinkedRecording(
        @Header("Authorization") authorization: String,
        @Path("deviceId") deviceId: String,
    ): OkResponse

    @POST("recordings/device/playback-url")
    suspend fun linkedRecordingPlayback(
        @Header("Authorization") authorization: String,
        @Body body: RecordingPlaybackRequest,
    ): RecordingPlaybackResponse

    @PATCH("devices/me/status")
    suspend fun updateStatus(
        @Header("Authorization") authorization: String,
        @Body body: DeviceStatusUpdate,
    ): DeviceStatusResponse

    @PUT("device-push-tokens")
    suspend fun registerPushToken(
        @Header("Authorization") authorization: String,
        @Body body: RegisterPushTokenRequest,
    ): OkResponse

    @POST("streaming/publisher-token")
    suspend fun publisherToken(
        @Header("Authorization") authorization: String,
    ): PublisherTokenResponse

    @GET("device-chats")
    suspend fun chats(
        @Header("Authorization") authorization: String,
    ): List<ChatThreadDto>

    @POST("device-chats/support")
    suspend fun openSupportChat(
        @Header("Authorization") authorization: String,
    ): ChatThreadDto

    @GET("device-chats/support/summary")
    suspend fun supportSummary(
        @Header("Authorization") authorization: String,
    ): SupportSummaryDto

    @GET("device-chats/{id}")
    suspend fun chatThread(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
    ): ChatThreadDto

    @GET("device-chats/{id}/messages")
    suspend fun chatMessages(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Query("cursor") cursor: String? = null,
        @Query("take") take: Int? = null,
    ): ChatMessagesPage

    @GET("device-chats/{id}/search")
    suspend fun searchChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Query("q") query: String,
    ): ChatSearchPage

    @GET("device-chats/{id}/media")
    suspend fun chatMedia(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Query("kind") kind: String,
    ): ChatMediaPage

    @POST("device-chats/{id}/messages")
    suspend fun sendChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Body body: SendChatRequest,
    ): ChatMessageDto

    @PATCH("device-chats/{id}/messages/{messageId}")
    suspend fun editChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Path("messageId") messageId: String,
        @Body body: EditChatRequest,
    ): ChatMessageDto

    @DELETE("device-chats/{id}/messages/{messageId}")
    suspend fun deleteChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Path("messageId") messageId: String,
        @Query("forEveryone") forEveryone: Boolean,
    ): OkResponse

    @POST("device-chats/{id}/messages/{messageId}/react")
    suspend fun reactChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Path("messageId") messageId: String,
        @Body body: ReactChatRequest,
    ): ChatMessageDto

    @POST("device-chats/{id}/read")
    suspend fun readChat(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
    ): OkResponse

    @POST("device-chats/{id}/uploads")
    suspend fun initUpload(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Body body: InitUploadRequest,
    ): InitUploadResponse

    @DELETE("device-chats/{id}/uploads/{uploadId}")
    suspend fun cancelUpload(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
        @Path("uploadId") uploadId: String,
    ): OkResponse

    @GET("device-subscriptions/me")
    suspend fun subscription(
        @Header("Authorization") authorization: String,
    ): SubscriptionDto

    @POST("device-subscriptions/invoices")
    suspend fun createPaymentInvoice(
        @Header("Authorization") authorization: String,
        @Body body: PurchasePlanRequest,
    ): PaymentInvoiceDto

    @GET("device-subscriptions/invoices/{id}")
    suspend fun paymentInvoice(
        @Header("Authorization") authorization: String,
        @Path("id") id: String,
    ): PaymentInvoiceDto
}

File: apps/user-web/src/lib/device-api.ts
import { api, fileUrl } from "./api";
import type {
  ChatMediaPage,
  ChatMessageDto,
  ChatMessagesPage,
  ChatSearchPage,
  ChatThreadDto,
  DeviceMeResponse,
  DeviceRecordingDto,
  InitUploadResponse,
  LinkDeviceResponse,
  LinkedDeviceDto,
  PairingCodeResponse,
  PairResponse,
  PairStatusResponse,
  PaymentInvoiceDto,
  PlaybackUrlResponse,
  SubscriptionDto,
  SupportSummaryDto,
  ViewerTokenResponse,
} from "./types";

export const deviceApi = {
  pairStatus: (phone: string) =>
    api.get<PairStatusResponse>(
      `/devices/pair-status?phone=${encodeURIComponent(phone)}`,
      { auth: false },
    ),

  pair: (body: {
    phone: string;
    password: string;
    name?: string;
    code?: string;
    appVersion?: string;
    deviceModel?: string;
  }) =>
    api.post<PairResponse>("/devices/pair", body, { auth: false }),

  guestSupport: (body: {
    installId: string;
    name?: string;
    appVersion?: string;
    deviceModel?: string;
    installSignals?: string[];
  }) =>
    api.post<PairResponse & { guest?: boolean }>(
      "/devices/guest-support",
      body,
      { auth: false },
    ),

  me: () => api.get<DeviceMeResponse>("/devices/me"),

  updateProfile: (body: { name?: string; phone?: string }) =>
    api.patch<DeviceMeResponse>("/devices/me", body),

  changePassword: (body: { currentPassword: string; newPassword: string }) =>
    api.patch<{ ok?: boolean }>("/devices/me/password", body),

  uploadAvatar: (imageBase64: string) =>
    api.post<DeviceMeResponse>("/devices/me/avatar", { imageBase64 }),

  deleteAvatar: () => api.delete<{ ok?: boolean }>("/devices/me/avatar"),

  linked: () => api.get<LinkedDeviceDto[]>("/devices/me/linked"),

  unlink: (id: string) =>
    api.delete<{ ok?: boolean }>(`/devices/me/linked/${id}`),

  createPairingCode: () =>
    api.post<PairingCodeResponse>("/devices/me/pairing-codes", {}),

  linkDevice: (code: string) =>
    api.post<LinkDeviceResponse>("/devices/me/link", { code }),

  setCameraFacing: (id: string, facing: "FRONT" | "BACK") =>
    api.post<{ ok?: boolean }>(`/devices/me/linked/${id}/camera`, { facing }),

  chats: () => api.get<ChatThreadDto[]>("/device-chats"),

  openSupport: () => api.post<ChatThreadDto>("/device-chats/support"),

  supportSummary: () =>
    api.get<SupportSummaryDto>("/device-chats/support/summary"),

  thread: (id: string) => api.get<ChatThreadDto>(`/device-chats/${id}`),

  messages: (id: string, cursor?: string | null, take = 50) => {
    const q = new URLSearchParams({ take: String(take) });
    if (cursor) q.set("cursor", cursor);
    return api.get<ChatMessagesPage>(`/device-chats/${id}/messages?${q}`);
  },

  searchMessages: (id: string, q: string) =>
    api.get<ChatSearchPage>(
      `/device-chats/${id}/search?q=${encodeURIComponent(q)}`,
    ),

  media: (id: string, kind: string) =>
    api.get<ChatMediaPage>(
      `/device-chats/${id}/media?kind=${encodeURIComponent(kind)}`,
    ),

  sendMessage: (
    id: string,
    body: {
      text: string;
      replyToId?: string;
      clientId?: string;
      forwardedFromId?: string;
    },
  ) => api.post<ChatMessageDto>(`/device-chats/${id}/messages`, body),

  editMessage: (threadId: string, messageId: string, text: string) =>
    api.patch<ChatMessageDto>(
      `/device-chats/${threadId}/messages/${messageId}`,
      { text },
    ),

  deleteMessage: (
    threadId: string,
    messageId: string,
    forEveryone = false,
  ) =>
    api.delete<ChatMessageDto>(
      `/device-chats/${threadId}/messages/${messageId}?forEveryone=${forEveryone}`,
    ),

  react: (threadId: string, messageId: string, emoji: string) =>
    api.post<ChatMessageDto>(
      `/device-chats/${threadId}/messages/${messageId}/react`,
      { emoji },
    ),

  markRead: (threadId: string) =>
    api.post<{ ok?: boolean }>(`/device-chats/${threadId}/read`),

  cancelUpload: (threadId: string, uploadId: string) =>
    api.delete<{ ok?: boolean }>(
      `/device-chats/${threadId}/uploads/${uploadId}`,
    ),

  subscription: () =>
    api.get<SubscriptionDto>("/device-subscriptions/me"),

  createInvoice: (plan: "PRO" | "PRO_PLUS") =>
    api.post<PaymentInvoiceDto>("/device-subscriptions/invoices", { plan }),

  invoice: (id: string) =>
    api.get<PaymentInvoiceDto>(`/device-subscriptions/invoices/${id}`),

  viewerToken: (deviceId: string) =>
    api.post<ViewerTokenResponse>(
      `/streaming/devices/${deviceId}/device-viewer-token`,
    ),

  deviceRecordings: async (deviceId: string) => {
    const res = await api.get<{ items?: DeviceRecordingDto[] } | DeviceRecordingDto[]>(
      `/recordings/device/${deviceId}`,
    );
    return Array.isArray(res) ? res : (res.items ?? []);
  },

  startRecording: (deviceId: string) =>
    api.post<{ ok?: boolean }>(`/recordings/device/${deviceId}/start`, {}),

  playbackUrl: (body: { id: string }) =>
    api.post<PlaybackUrlResponse>("/recordings/device/playback-url", body),
};

export type UploadOpts = {
  messageType: string;
  replyToId?: string;
  clientId?: string;
  text?: string;
  albumId?: string;
  durationMs?: number;
  width?: number;
  height?: number;
  waveformJson?: string;
  onProgress?: (ratio: number) => void;
};

export async function uploadChatFile(
  threadId: string,
  file: File,
  opts: UploadOpts,
): Promise<ChatMessageDto> {
  const init = await api.post<InitUploadResponse>(
    `/device-chats/${threadId}/uploads`,
    {
      fileName: file.name,
      mimeType: file.type || "application/octet-stream",
      fileSize: file.size,
      messageType: opts.messageType,
      replyToId: opts.replyToId,
      clientId: opts.clientId,
      text: opts.text,
      albumId: opts.albumId,
      durationMs: opts.durationMs,
      width: opts.width,
      height: opts.height,
      waveformJson: opts.waveformJson,
    },
  );

  const chunkSize = init.chunkSize || 262144;
  const buffer = await file.arrayBuffer();
  const total = Math.ceil(buffer.byteLength / chunkSize) || 1;

  for (let i = 0; i < total; i++) {
    const start = i * chunkSize;
    const end = Math.min(start + chunkSize, buffer.byteLength);
    const chunk = buffer.slice(start, end);
    await api.putBinary(
      `/device-chats/${threadId}/uploads/${init.uploadId}/chunks/${i}`,
      chunk,
    );
    opts.onProgress?.((i + 1) / total);
  }

  return api.post<ChatMessageDto>(
    `/device-chats/${threadId}/uploads/${init.uploadId}/complete`,
  );
}

export function authFileUrl(
  threadId: string,
  messageId: string,
  thumb = false,
): string {
  return fileUrl(threadId, messageId, thumb);
}
